### PR TITLE
Add Udev Rules to README

### PR DESCRIPTION
**Pull request by lujinhao**

### CODE REVIEW

**Comment by lujinhao** [comment]
@garethellis0 I have added the udev rules under the README Arduino section.

File: README.md [hunk 1]
```diff
@@ -182,6 +182,38 @@ some_ros_package
 - When developing the firmware/Arduino parts of the software, we've made a complete arduino workspace in `src/firmware`. This way you don't need to worry about downloading the libraries yourself!
 - In order to use this, go to your Arduino IDE's Preferences dialog box and use `/your/path/to/Snowflake/src/firmware` as your sketchbook directory. Open arduino sketches in the workspace and they will work!
 
+### Udev Rules
+
+Udev provides a userspace solution for a dynamic /dev directory, with persistent device naming.
```
Comment (garethellis0) [inline README.md]: Practically speaking, what does this allow us to do? Why is it important?

File: README.md [hunk 2]
```diff
@@ -182,6 +182,38 @@ some_ros_package
...
+
+- Rule Writing Guideline
+	- Creat file: /etc/udev/rules.d/10-local.rules (where xx represent the priority; Here 10 is higher than 50, the default)
```
Comment (garethellis0) [inline README.md]: Spelling: `Create`

Comment (garethellis0) [reply README.md]: We should probably make sure this is Snowbots specific, so we can overwrite it if needed (maybe `10-snowbots.rules`)

**Comment by lujinhao** [comment]
changes made!

File: src/firmware/README.md [hunk 3]
```diff
@@ -1,12 +1,34 @@
 # Fimware
 
-### gps_firmware
+### Udev Rules
 
-- Designed to allow pass lat and lon coordinates from a [Adafruit Ultimate GPS Breakout v3](https://www.adafruit.com/product/746) to a PC over serial
-- Sends messages in the form `GPS,lat,lon,has_gps_fix`, where `has_gps_fix` is a 1 or 0 representing whether or not the GPS has a fix
+Udev assigns the names specified in the rule file to the devices, which helps matching dev files with devices.
```
Comment (garethellis0) [inline src/firmware/README.md]: We should add something here explaining things here as if the person reading it has no knowledge of dev files or USB devices. My suggestion would be adding something like:
```
More simply, udev rules allow us to guarantee that no matter what order we plug USB devices in, they will always be accessible at the same path in software.
```

File: src/firmware/README.md [hunk 4]
```diff
@@ -1,12 +1,34 @@
...
 
-char to send | what it does
--------------| ------------
-`d` | toggles debug mode. in debug mode, the gps will send messages over serial as it receives them
-`r` | requests the most recent gps message (it will be returned over serial)
+- Rule Writing Guideline
+	- Create file: `/etc/udev/rules.d/10-snowbots.rules` (where xx represent the priority; Here 10 is higher than 50, the default)
```
Comment (garethellis0) [inline src/firmware/README.md]: Since we've already created the file, can we remove this step?

File: get_started.sh [hunk 5]
```diff
@@ -106,6 +106,12 @@ sudo ln -s -f /usr/share/clion*/bin/clion.sh /usr/local/bin/clion
 cd $DIR
 ./install_dependencies.sh
 
+##############################
+# Setup Snowbots Udev Rules  #
+##############################
+cd $DIR
+./setup_udev_rules.sh
```
Comment (garethellis0) [inline get_started.sh]: Doesn't look like `setup_udev_rules.sh` is part of this pull request?

Comment (garethellis0) [reply get_started.sh]: Also, we should put the script with the rest of the firmware stuff in `src/firmware`. We'd also have to change the path here to: `./src/firmware/setup_udev_rules.sh`

File: src/firmware/README.md [hunk 6]
```diff
@@ -1,12 +1,34 @@
...
+	- Write rules with match keys (helps indentify devices) and assignment keys (assigns names/symlinks to identified devices)
+	    - common match keys: 
+	    	KERNEL - match against the kernel name for the device
+	    	SUBSYSTEM - match against the subsystem of the device
+	    	DRIVER - match against the name of the driver backing the device
+	    - common assignment keys:
+	    	NAME - the name that shall be used for the device node
+	    	SYMLINK - a list of symbolic links which act as alternative names for the device node
+	    	MODE - define Unix permissions, by default 0666 (read/write to owner group)
+	    	GROUP - define the Unix group ownship of the device node
```
Comment (garethellis0) [inline src/firmware/README.md]: Should a few more steps here:
```
- add your new rule to `10-snowbots.rules`
- run `setup_udev_rules.sh` to apply your new rules
```

File: src/firmware/README.md [hunk 7]
```diff
@@ -1,12 +1,34 @@
 # Fimware
 
-### gps_firmware
+### Udev Rules
```
Comment (garethellis0) [inline src/firmware/README.md]: We should add an entry for this in the table of contents at the top of the `README` under `Arduino Development`. (We may also want to consider renaming `Arduino Development` to something like `Firmware Development`)

File: src/firmware/README.md [hunk 8]
```diff
@@ -1,12 +1,34 @@
...
-- Sends messages in the form `GPS,lat,lon,has_gps_fix`, where `has_gps_fix` is a 1 or 0 representing whether or not the GPS has a fix
+Udev assigns the names specified in the rule file to the devices, which helps matching dev files with devices.
 
-char to send | what it does
--------------| ------------
-`d` | toggles debug mode. in debug mode, the gps will send messages over serial as it receives them
-`r` | requests the most recent gps message (it will be returned over serial)
+- Rule Writing Guideline
+	- Create file: `/etc/udev/rules.d/10-snowbots.rules` (where xx represent the priority; Here 10 is higher than 50, the default)
+	- Write rules with match keys (helps indentify devices) and assignment keys (assigns names/symlinks to identified devices)
```
Comment (garethellis0) [inline src/firmware/README.md]: `indentify` -> `identify`

File: src/firmware/README.md (hunk 7 above)
Comment (lujinhao) [reply src/firmware/README.md]: I think "Udev Rules" entry is in the file under the `firmware folder` which is separate from the README in the main directory. In this case, there is no `Arduino Development` heading.

File: src/firmware/README.md (hunk 4 above)
Comment (lujinhao) [reply src/firmware/README.md]: Right. I'm wondering if this line can still be kept with a comment below saying this step is already done automatically in the first time setup. Creating the rule file is the first step of the udev rules setup. The complete udev rules writing guideline allows other members to make some changes in the bash file for the udev rules setup later.

**Comment by lujinhao** [comment]
changes made!

File: src/firmware/10-snowbots.rules [hunk 9]
```diff
@@ -0,0 +1,5 @@
+# Snowbots udev Rules
```
[the remark this answers is not in the document]
Comment (lujinhao) [reply src/firmware/10-snowbots.rules]: The serial attrs are made when the devices were manufactured. What I did was checking them on my laptop and manually note the serial numbers down. 

Based on the test last time, I believe the Udev rules will work unless something goes wrong with the `udev setup` bash file.

File: src/firmware/README.md (hunk 7 above)
Comment (garethellis0) [reply src/firmware/README.md]: Good point! :+1:

File: src/firmware/README.md (hunk 4 above)
Comment (garethellis0) [reply src/firmware/README.md]: I think that if the bash file is sufficiently commented, this should be enough documentation for people to understand how we're creating the udev rules. The README should just detail what people will need to know in order to write and update them.

**Comment by garethellis0** [comment]
@lujinhao - I think you missed a couple of comments on your first pass through. Could you please go through all my comments above and reply to each of them (even if it's just a simple :+1: ) so I know you've seen and dealt with them?

File: src/firmware/README.md (hunk 6 above)
Comment (lujinhao) [reply src/firmware/README.md]: changes made

File: src/firmware/README.md (hunk 8 above)
Comment (lujinhao) [reply src/firmware/README.md]: changes made

File: src/firmware/README.md (hunk 3 above)
Comment (lujinhao) [reply src/firmware/README.md]: changes made

File: README.md (hunk 2 above)
Comment (lujinhao) [reply README.md]: changes made

File: README.md (hunk 1 above)
Comment (lujinhao) [reply README.md]: changes made

File: src/firmware/README.md (hunk 4 above)
Comment (lujinhao) [reply src/firmware/README.md]: Ok, I'll delete the first step and push it again.

**Comment by lujinhao** [comment]
@garethellis0 - I have replied to all the comments above and made changes accordingly 👍

**Comment by lujinhao** [comment]
Cheers!

**Comment by lujinhao** [comment]
I'm not sure how to merge the forked repo to the original repo. I might need to get some help from you this Saturday to merge it. Thanks.

**Comment by garethellis0** [comment]
Hey @lujinhao, there should be a `Squash and merge` button on this pull request, can you see it?
screenshot from 2017-10-11 03-05-01

**Comment by lujinhao** [comment]
@garethellis0 - I'm afraid that it doesn't show up on my side.
image

**Comment by garethellis0** [comment]
Hmm. Odd. Not a big deal, so I'll merge this for now.

**Pull request merged**